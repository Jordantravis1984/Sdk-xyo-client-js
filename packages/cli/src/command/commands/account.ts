import { parse } from 'path'
import { ArgumentsCamelCase, Argv, CommandBuilder, CommandModule } from 'yargs'

import { opts } from '../requireDirectoryOptions'
// eslint-disable-next-line @typescript-eslint/ban-types
type Arguments = {}

export const aliases: ReadonlyArray<string> = []
export const builder: CommandBuilder = (yargs: Argv) =>
  yargs.usage('Usage: $0 account <command> [Options]').commandDir(parse(__filename).name, opts).demandCommand().version(false)
export const command = 'account <command> [Options]'
export const deprecated = false
export const describe = 'Create & manage your XYO account'
export const handler = function (_argv: ArgumentsCamelCase<Arguments>) {
  // do something with argv.
  console.log('handler')
}

const mod: CommandModule = {
  aliases,
  command,
  deprecated,
  describe,
  handler,
}

// eslint-disable-next-line import/no-default-export
export default mod
