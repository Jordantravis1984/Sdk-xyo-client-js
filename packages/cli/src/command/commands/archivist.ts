import { parse } from 'path'
import { ArgumentsCamelCase, Argv, CommandBuilder, CommandModule } from 'yargs'

import { opts } from '../requireDirectoryOptions'

// eslint-disable-next-line @typescript-eslint/ban-types
type Arguments = {}

export const aliases: ReadonlyArray<string> = []
export const builder: CommandBuilder = (yargs: Argv) =>
  yargs.usage('Usage: $0 archivist <query> <address> [Options]').commandDir(parse(__filename).name, opts).demandCommand().version(false)
export const command = 'archivist'
export const deprecated = false
export const describe = 'Issue queries against an XYO archivist'
export const handler = function (argv: ArgumentsCamelCase<Arguments>) {
  console.log(command)
  console.log(argv)
}

const mod: CommandModule = {
  aliases,
  command,
  deprecated,
  describe,
  handler,
}

// eslint-disable-next-line import/no-default-export
export default mod
