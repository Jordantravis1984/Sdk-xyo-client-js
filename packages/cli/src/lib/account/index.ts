export * from './files'
export * from './getAccount'
export * from './getAccountInfo'
