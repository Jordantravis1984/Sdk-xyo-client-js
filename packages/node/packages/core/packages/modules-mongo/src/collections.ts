// TODO: By DB
export const COLLECTIONS = {
  ArchiveKeys: 'archive_keys' as const,
  Archives: 'archives' as const,
  ArchivistStats: 'archivist_stats' as const,
  BoundWitnesses: 'bound_witnesses' as const,
  Payloads: 'payloads' as const,
  Users: 'users' as const,
}
