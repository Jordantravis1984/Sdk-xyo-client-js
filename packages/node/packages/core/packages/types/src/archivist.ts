import { ModuleList } from './ModuleList'

export const ARCHIVIST_TYPES: ModuleList = {
  Archivist: Symbol('Archivist'),
  BoundWitnessArchivist: Symbol('BoundWitnessArchivist'),
  PayloadArchivist: Symbol('PayloadArchivist'),
  UserArchivist: Symbol('UserArchivist'),
}
