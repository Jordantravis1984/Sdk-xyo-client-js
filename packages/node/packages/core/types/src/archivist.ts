import { ModuleList } from './ModuleList'

export const ARCHIVIST_TYPES: ModuleList = {
  ArchiveArchivist: Symbol('ArchiveArchivist'),
  ArchiveBoundWitnessArchivistFactory: Symbol('ArchiveBoundWitnessArchivistFactory'),
  ArchiveKeyRepository: Symbol('ArchiveKeyRepository'),
  ArchivePayloadArchivistFactory: Symbol('ArchivePayloadArchivistFactory'),
  ArchivePermissionsArchivistFactory: Symbol('ArchivePermissionsArchivistFactory'),
  BoundWitnessArchivist: Symbol('BoundWitnessArchivist'),
  PayloadArchivist: Symbol('PayloadArchivist'),
  UserArchivist: Symbol('UserArchivist'),
  WitnessedPayloadArchivist: Symbol('WitnessedPayloadArchivist'),
}
