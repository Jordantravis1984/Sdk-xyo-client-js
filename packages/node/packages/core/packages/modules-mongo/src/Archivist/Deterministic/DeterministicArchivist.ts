import { assertEx } from '@xylabs/assert'
import { exists } from '@xylabs/exists'
import { fulfilledValues } from '@xylabs/promise'
import { Account } from '@xyo-network/account'
import {
  AbstractArchivist,
  ArchivistConfig,
  ArchivistConfigSchema,
  ArchivistFindQuery,
  ArchivistFindQuerySchema,
  ArchivistGetQuery,
  ArchivistGetQuerySchema,
  ArchivistInsertQuery,
  ArchivistInsertQuerySchema,
  ArchivistParams,
  ArchivistQuery,
} from '@xyo-network/archivist'
import { XyoBoundWitness, XyoBoundWitnessSchema } from '@xyo-network/boundwitness-model'
import { BoundWitnessWrapper } from '@xyo-network/boundwitness-wrapper'
import {
  AnyConfigSchema,
  ModuleConfig,
  ModuleEventData,
  ModuleQueryResult,
  QueryBoundWitnessWrapper,
  XyoErrorBuilder,
  XyoQueryBoundWitness,
} from '@xyo-network/module'
import { XyoBoundWitnessWithMeta, XyoPayloadWithMeta, XyoPayloadWithPartialMeta } from '@xyo-network/node-core-model'
import { PayloadFindFilter, XyoPayload } from '@xyo-network/payload-model'
import { PayloadWrapper } from '@xyo-network/payload-wrapper'
import { BaseMongoSdk } from '@xyo-network/sdk-xyo-mongo-js'
import { SortDirection } from 'mongodb'

import { DefaultMaxTimeMS } from '../../defaults'
import {
  BoundWitnessesFilter,
  getArchive,
  getFilter,
  getLimit,
  getPayloadSchemas,
  PayloadsFilter,
  shouldFindBoundWitnesses,
  shouldFindPayloads,
} from './QueryHelpers'
import { validByType } from './validByType'

export type MongoDBDeterministicArchivistParams = ArchivistParams<
  AnyConfigSchema<ArchivistConfig>,
  ModuleEventData,
  {
    boundWitnesses: BaseMongoSdk<XyoBoundWitnessWithMeta>
    payloads: BaseMongoSdk<XyoPayloadWithMeta>
  }
>

const toBoundWitnessWithMeta = (wrapper: BoundWitnessWrapper | QueryBoundWitnessWrapper, archive: string): XyoBoundWitnessWithMeta => {
  const bw = wrapper.boundwitness as XyoBoundWitness
  return { ...bw, _archive: archive, _hash: wrapper.hash, _timestamp: Date.now() }
}

const toReturnValue = (value: XyoPayload | XyoBoundWitness): XyoPayload => {
  const _signatures = (value as XyoBoundWitness)?._signatures
  if (_signatures) {
    return { ...PayloadWrapper.parse(value).body, _signatures } as XyoPayload
  } else {
    return { ...PayloadWrapper.parse(value).body }
  }
}

const toPayloadWithMeta = (wrapper: PayloadWrapper, archive: string): XyoPayloadWithMeta => {
  return { ...wrapper.payload, _archive: archive, _hash: wrapper.hash, _timestamp: Date.now() }
}

const searchDepthLimit = 50

export class MongoDBDeterministicArchivist<
  TParams extends MongoDBDeterministicArchivistParams = MongoDBDeterministicArchivistParams,
> extends AbstractArchivist<TParams> {
  static override configSchema = ArchivistConfigSchema

  get boundWitnesses() {
    return this.params.boundWitnesses
  }

  get payloads() {
    return this.params.payloads
  }

  override get queries(): string[] {
    return [ArchivistFindQuerySchema, ArchivistInsertQuerySchema, ...super.queries]
  }

  override find(_filter?: PayloadFindFilter): Promise<XyoPayload[]> {
    throw new Error('find method must be called via query')
  }

  override get(_items: string[]): Promise<XyoPayload[]> {
    throw new Error('get method must be called via query')
  }

  insert(_items: XyoPayload[]): Promise<XyoBoundWitness[]> {
    throw new Error('insert method must be called via query')
  }

  override async query<T extends XyoQueryBoundWitness = XyoQueryBoundWitness, TConfig extends ModuleConfig = ModuleConfig>(
    query: T,
    payloads?: XyoPayload[],
    queryConfig?: TConfig,
  ): Promise<ModuleQueryResult> {
    const wrapper = QueryBoundWitnessWrapper.parseQuery<ArchivistQuery>(query, payloads)
    const typedQuery = wrapper.query.payload
    assertEx(this.queryable(query, payloads, queryConfig))
    const resultPayloads: XyoPayload[] = []
    // TODO: Use new Account once we mock Account.new in Jest
    const queryAccount = Account.random()
    // const queryAccount = new Account()
    try {
      switch (typedQuery.schema) {
        case ArchivistFindQuerySchema: {
          resultPayloads.push(...(await this.findInternal(wrapper, typedQuery)))
          break
        }
        case ArchivistGetQuerySchema: {
          resultPayloads.push(...(await this.getInternal(wrapper, typedQuery)))
          break
        }
        case ArchivistInsertQuerySchema: {
          resultPayloads.push(...(await this.insertInternal(wrapper, typedQuery)))
          break
        }
        default:
          return super.query(query, payloads)
      }
    } catch (ex) {
      const error = ex as Error
      resultPayloads.push(new XyoErrorBuilder([wrapper.hash], error.message).build())
    }
    return this.bindResult(resultPayloads, queryAccount)
  }

  protected async findBoundWitness(
    filter: BoundWitnessesFilter,
    sort: { [key: string]: SortDirection } = { _timestamp: -1 },
  ): Promise<XyoBoundWitnessWithMeta | undefined> {
    return (await (await this.boundWitnesses.find(filter)).sort(sort).limit(1).maxTimeMS(DefaultMaxTimeMS).toArray()).pop()
  }

  protected async findInternal(wrapper: QueryBoundWitnessWrapper<ArchivistQuery>, typedQuery: ArchivistFindQuery): Promise<XyoPayload[]> {
    const limit = getLimit(typedQuery)
    const [bwFilter, payloadFilter] = getFilter(wrapper, typedQuery)
    const resultPayloads: (XyoBoundWitnessWithMeta | XyoPayloadWithMeta)[] = []
    const findBWs = shouldFindBoundWitnesses(typedQuery)
    const findPayloads = shouldFindPayloads(typedQuery)
    const payloadSchemas = getPayloadSchemas(typedQuery)
    let currentBw: XyoBoundWitnessWithMeta | undefined
    let nextHash: string | null | undefined = undefined
    for (let searchDepth = 0; searchDepth < searchDepthLimit; searchDepth++) {
      if (resultPayloads.length >= limit) break
      currentBw = await this.findBoundWitness(bwFilter)
      if (!currentBw) break
      if (findBWs) resultPayloads.push(currentBw)
      if (findPayloads) {
        const _archive = getArchive(currentBw)
        const payloadHashes = payloadSchemas.length
          ? currentBw?.payload_schemas.reduce((schemas, schema, idx) => {
              if (payloadSchemas.includes(schema) && currentBw?.payload_hashes[idx]) schemas.push(currentBw.payload_hashes[idx])
              return schemas
            }, [] as string[])
          : currentBw.payload_hashes
        const payloads = (
          await Promise.all(
            payloadHashes.map((_hash) => {
              const schema = currentBw?.payload_schemas[currentBw.payload_hashes.indexOf(_hash)]
              return schema?.startsWith(XyoBoundWitnessSchema)
                ? this.findBoundWitness({ _archive, _hash })
                : this.findPayload({ ...payloadFilter, _archive, _hash })
            }),
          )
        ).filter(exists)
        for (let p = 0; p < payloads.length; p++) {
          if (resultPayloads.length >= limit) break
          resultPayloads.push(payloads[p])
        }
      }
      nextHash = currentBw?.previous_hashes?.at(-1)
      if (!nextHash) break
      bwFilter._hash = nextHash
    }
    return resultPayloads.map(toReturnValue)
  }

  protected async findPayload(filter: PayloadsFilter): Promise<XyoPayloadWithMeta | undefined> {
    const { _archive, order, offset, schema, _hash } = filter as PayloadFindFilter & PayloadsFilter
    const sort: { [key: string]: SortDirection } = { _timestamp: order === 'asc' ? 1 : -1 }
    const parsedTimestamp = offset ? parseInt(`${offset}`) : order === 'asc' ? 0 : Date.now()
    const _timestamp = order === 'asc' ? { $gte: parsedTimestamp } : { $lte: parsedTimestamp }
    const find: PayloadsFilter = { _archive, _timestamp }
    if (schema) find.schema = schema
    if (_hash) find._hash = _hash
    const result = await (await this.payloads.find(find)).limit(1).sort(sort).maxTimeMS(DefaultMaxTimeMS).toArray()
    return result.pop()
  }

  protected async getInternal(wrapper: QueryBoundWitnessWrapper<ArchivistQuery>, typedQuery: ArchivistGetQuery): Promise<XyoPayload[]> {
    const hashes = typedQuery.hashes
    const payloads = hashes.map((_hash) => this.payloads.findOne({ _hash }))
    const bws = hashes.map((_hash) => this.boundWitnesses.findOne({ _hash }))
    const gets = await Promise.allSettled([payloads, bws].flat())
    const succeeded = gets.reduce<(XyoPayloadWithPartialMeta | null)[]>(fulfilledValues, [])
    return succeeded.filter(exists).map(toReturnValue)
  }

  protected async insertInternal(wrapper: QueryBoundWitnessWrapper<ArchivistQuery>, _typedQuery: ArchivistInsertQuery): Promise<XyoBoundWitness[]> {
    const archive = this.address
    const toStore = [wrapper.boundwitness, ...wrapper.payloadsArray.map((p) => p.payload)]
    const [bw, p] = toStore.reduce(validByType, [[], []])
    const boundWitnesses = bw.map((x) => toBoundWitnessWithMeta(x, archive))
    const payloads = p.map((x) => toPayloadWithMeta(x, archive))
    if (boundWitnesses.length) {
      const boundWitnessesResult = await this.boundWitnesses.insertMany(boundWitnesses)
      if (!boundWitnessesResult.acknowledged || boundWitnessesResult.insertedCount !== boundWitnesses.length)
        throw new Error('MongoDBDeterministicArchivist: Error inserting BoundWitnesses')
    }
    if (payloads.length) {
      const payloadsResult = await this.payloads.insertMany(payloads)
      if (!payloadsResult.acknowledged || payloadsResult.insertedCount !== payloads.length)
        throw new Error('MongoDBDeterministicArchivist: Error inserting Payloads')
    }
    const result = await this.bindResult([wrapper.boundwitness, ...wrapper.payloadsArray.map((p) => p.payload)])
    return [result[0]]
  }
}
