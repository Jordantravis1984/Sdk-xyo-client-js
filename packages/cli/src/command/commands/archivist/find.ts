import { ArgumentsCamelCase, Argv, CommandBuilder, CommandModule } from 'yargs'

import { printLine } from '../../../lib'

// eslint-disable-next-line @typescript-eslint/ban-types
type Arguments = {}

export const aliases: ReadonlyArray<string> = []
export const builder: CommandBuilder = (yargs: Argv) =>
  yargs
    .usage('Usage: $0 archivist find <address> <filter>')
    .positional('address', { demandOption: true, type: 'string' })
    .positional('filter', { demandOption: true, type: 'string' })
    .version(false)

export const command = 'find <address> <filter>'
export const deprecated = false
export const describe = 'Find payload(s) in the Archivist matching the supplied filter'
export const handler = function (argv: ArgumentsCamelCase<Arguments>) {
  printLine(JSON.stringify(command))
  printLine(JSON.stringify(argv))
}

const mod: CommandModule = {
  aliases,
  command,
  deprecated,
  describe,
  handler,
}

// eslint-disable-next-line import/no-default-export
export default mod
