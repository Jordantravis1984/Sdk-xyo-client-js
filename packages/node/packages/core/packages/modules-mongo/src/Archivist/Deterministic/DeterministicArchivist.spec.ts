import { assertEx } from '@xylabs/assert'
import { Account } from '@xyo-network/account'
import { ArchivistWrapper } from '@xyo-network/archivist'
import { BoundWitnessBuilder } from '@xyo-network/boundwitness-builder'
import { XyoBoundWitnessSchema } from '@xyo-network/boundwitness-model'
import { BoundWitnessWrapper } from '@xyo-network/boundwitness-wrapper'
import { AbstractModuleConfigSchema } from '@xyo-network/module-model'
import { XyoBoundWitnessWithMeta, XyoPayloadWithMeta } from '@xyo-network/node-core-model'
import { PayloadWrapper } from '@xyo-network/payload-wrapper'
import { BaseMongoSdk, BaseMongoSdkConfig } from '@xyo-network/sdk-xyo-mongo-js'
import { MongoMemoryServer } from 'mongodb-memory-server'

import { COLLECTIONS } from '../../collections'
import { MongoDBDeterministicArchivist } from './DeterministicArchivist'

describe('DeterministicArchivist', () => {
  const boundWitnessesConfig: BaseMongoSdkConfig = { collection: COLLECTIONS.BoundWitnesses }
  const payloadsConfig: BaseMongoSdkConfig = { collection: COLLECTIONS.Payloads }
  const server = new MongoMemoryServer()
  let account: Account = Account.random()
  let archivist: ArchivistWrapper
  let nonce = Date.now()
  beforeAll(async () => {
    await server.start()
    const uri = server.getUri()
    boundWitnessesConfig.dbConnectionString = uri
    payloadsConfig.dbConnectionString = uri
  })
  afterAll(async () => {
    await server.stop()
  })
  beforeEach(async () => {
    nonce = Date.now()
    const boundWitnesses: BaseMongoSdk<XyoBoundWitnessWithMeta> = new BaseMongoSdk(boundWitnessesConfig)
    const payloads: BaseMongoSdk<XyoPayloadWithMeta> = new BaseMongoSdk(payloadsConfig)
    const module = await MongoDBDeterministicArchivist.create({ boundWitnesses, config: { schema: AbstractModuleConfigSchema }, payloads })
    account = Account.random()
    archivist = new ArchivistWrapper(module, account)
  })
  describe('discover', () => {
    it('discovers module', async () => {
      const result = await archivist.discover()
      expect(result).toBeArray()
      expect(result.length).toBeGreaterThan(0)
    })
  })
  describe('insert', () => {
    const payload1 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.debug' })
    const payload2 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.test' })
    it.each([
      ['inserts single payload', [payload1]],
      ['inserts multiple payloads', [payload1, payload2]],
    ])('%s', async (_title, payloads) => {
      const results = await archivist.insert(payloads.map((w) => w.payload))
      expect(results).toBeTruthy()
      expect(results).toBeArrayOfSize(2)
      const [boundResult, transactionResults] = results
      expect(boundResult.addresses).toContain(archivist.address)
      expect(transactionResults.addresses).toContain(account.public.address.hex)
      expect(transactionResults.payload_hashes).toBeArrayOfSize(payloads.length + 1)
      payloads.forEach((p) => {
        expect(transactionResults.payload_hashes).toInclude(p.hash)
      })
    })
  })
  describe('get', () => {
    const payload1 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.debug' })
    const payload2 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.test' })
    it.each([
      ['gets single payload', [payload1]],
      ['gets multiple payloads', [payload1, payload2]],
    ])('%s', async (_title, payloads) => {
      await archivist.insert(payloads.map((w) => w.payload))
      const results = await archivist.get(payloads.map((p) => p.hash))
      expect(results).toBeTruthy()
      expect(results).toBeArrayOfSize(payloads.length)
      const resultPayloads = results.map((result) => PayloadWrapper.parse(result))
      const resultHashes = resultPayloads.map((p) => p.hash)
      payloads.map((p) => {
        expect(resultHashes).toInclude(p.hash)
      })
    })
  })
  describe('find', () => {
    describe('with schema for BoundWitness', () => {
      const payload1 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.debug' })
      const payload2 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.test' })
      const schema = XyoBoundWitnessSchema
      it('finds single bw', async () => {
        const boundWitness1 = BoundWitnessWrapper.parse(new BoundWitnessBuilder().payload(payload1.payload).witness(account).build()[0])
        const boundWitnesses = [boundWitness1]
        await archivist.insert(boundWitnesses.map((bw) => bw.boundwitness))
        const limit = boundWitnesses.length
        const offset = assertEx(boundWitnesses.at(-1)?.hash)
        const results = await archivist.find({ limit, offset, schema })
        expect(results).toBeTruthy()
        expect(results).toBeArrayOfSize(boundWitnesses.length)
        const resultPayloads = results.map((result) => PayloadWrapper.parse(result))
        const resultHashes = resultPayloads.map((p) => p.hash)
        boundWitnesses.map((p) => {
          expect(resultHashes).toInclude(p.hash)
        })
      })
      it('finds multiple bws', async () => {
        const boundWitness1 = BoundWitnessWrapper.parse(new BoundWitnessBuilder().payload(payload1.payload).witness(account).build()[0])
        const boundWitness2 = BoundWitnessWrapper.parse(
          new BoundWitnessBuilder().payloads([payload1.payload, payload2.payload]).witness(account).build()[0],
        )
        const boundWitnesses = [boundWitness1, boundWitness2]
        await archivist.insert(boundWitnesses.map((bw) => bw.boundwitness))
        const limit = boundWitnesses.length
        const offset = assertEx(boundWitnesses.at(-1)?.hash)
        const results = await archivist.find({ limit, offset, schema })
        expect(results).toBeTruthy()
        expect(results).toBeArrayOfSize(boundWitnesses.length)
        const resultPayloads = results.map((result) => PayloadWrapper.parse(result))
        const resultHashes = resultPayloads.map((p) => p.hash)
        boundWitnesses.map((p) => {
          expect(resultHashes).toInclude(p.hash)
        })
      })
    })
    describe('with schema for Payload', () => {
      const schema = 'network.xyo.debug'
      const payload1 = PayloadWrapper.parse({ nonce, schema })
      const payload2 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.test' })
      it.each([
        ['finds single payload', [payload1]],
        ['finds multiple payloads', [payload1, payload2]],
      ])('%s', async (_title, payloads) => {
        for (let i = 0; i < payloads.length; i++) {
          const unique = payloads
            .map((w) => w.payload)
            .map((p) => {
              ;(p as unknown as { nonce: number }).nonce = Date.now()
              return p
            })
          await archivist.insert(unique)
        }
        const limit = payloads.length
        const results = await archivist.find({ limit, schema })
        expect(results).toBeTruthy()
        expect(results).toBeArrayOfSize(payloads.length)
        const resultPayloads = results.map((result) => PayloadWrapper.parse(result))
        resultPayloads.map((p) => {
          expect(p.schema).toBe(schema)
        })
      })
    })
    describe('with no schema', () => {
      const payload1 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.debug' })
      const payload2 = PayloadWrapper.parse({ nonce, schema: 'network.xyo.test' })
      it('finds address history', async () => {
        const payloads = [payload1, payload2]
        await Promise.all(payloads.map((payload) => archivist.insert([payload.payload])))
        const limit = payloads.length + 2
        const results = await archivist.find({ limit })
        expect(results).toBeTruthy()
        expect(results).toBeArrayOfSize(limit)
        const resultPayloads = results.map((result) => PayloadWrapper.parse(result))
        const resultHashes = resultPayloads.map((p) => p.hash)
        payloads.map((p) => {
          expect(resultHashes).toInclude(p.hash)
        })
      })
    })
  })
})
