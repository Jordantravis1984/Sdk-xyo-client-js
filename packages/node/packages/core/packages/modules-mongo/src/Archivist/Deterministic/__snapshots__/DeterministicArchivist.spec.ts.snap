// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeterministicArchivist find with no schema finds address history 1`] = `
[
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "1137543b5805a18e8f39cb7debe49ddf58a1de00c46ed7310baf7ebcfdf435a2",
      "aa36451ef0079af6cc2f86eaa05c7a93fb3c92098e18d0ce857e9e3072b32a85",
    ],
    "payload_schemas": [
      "network.xyo.debug",
      "network.xyo.test",
    ],
    "previous_hashes": [
      "d45ace1ac2f9e5d04dbf295850d47427ecd43f2e043c4a8326db2f8bf547dd38",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "nonce": 3,
    "schema": "network.xyo.debug",
  },
  {
    "nonce": 4,
    "schema": "network.xyo.test",
  },
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "1c9033e467d6877e2c84a70b29212a585bd6fd448bccfcc12adf0505c18984f6",
    ],
    "payload_schemas": [
      "network.xyo.test",
    ],
    "previous_hashes": [
      "d2fbc2df7cd956a2a9461792c9ffad4faf2ec290743832ff913582453e0dd320",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "nonce": 2,
    "schema": "network.xyo.test",
  },
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "nonce": 1,
    "schema": "network.xyo.debug",
  },
]
`;

exports[`DeterministicArchivist find with schema for BoundWitness finds multiple boundwitness 1`] = `
[
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "1137543b5805a18e8f39cb7debe49ddf58a1de00c46ed7310baf7ebcfdf435a2",
      "aa36451ef0079af6cc2f86eaa05c7a93fb3c92098e18d0ce857e9e3072b32a85",
    ],
    "payload_schemas": [
      "network.xyo.debug",
      "network.xyo.test",
    ],
    "previous_hashes": [
      "d45ace1ac2f9e5d04dbf295850d47427ecd43f2e043c4a8326db2f8bf547dd38",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "1c9033e467d6877e2c84a70b29212a585bd6fd448bccfcc12adf0505c18984f6",
    ],
    "payload_schemas": [
      "network.xyo.test",
    ],
    "previous_hashes": [
      "d2fbc2df7cd956a2a9461792c9ffad4faf2ec290743832ff913582453e0dd320",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist find with schema for BoundWitness finds single boundwitness 1`] = `
[
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist find with schema for Payload finds single payload 1`] = `
[
  {
    "nonce": 3,
    "schema": "network.xyo.debug",
  },
]
`;

exports[`DeterministicArchivist get gets multiple boundwitness 1`] = `
[
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "1c9033e467d6877e2c84a70b29212a585bd6fd448bccfcc12adf0505c18984f6",
    ],
    "payload_schemas": [
      "network.xyo.test",
    ],
    "previous_hashes": [
      "d2fbc2df7cd956a2a9461792c9ffad4faf2ec290743832ff913582453e0dd320",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "1137543b5805a18e8f39cb7debe49ddf58a1de00c46ed7310baf7ebcfdf435a2",
      "aa36451ef0079af6cc2f86eaa05c7a93fb3c92098e18d0ce857e9e3072b32a85",
    ],
    "payload_schemas": [
      "network.xyo.debug",
      "network.xyo.test",
    ],
    "previous_hashes": [
      "d45ace1ac2f9e5d04dbf295850d47427ecd43f2e043c4a8326db2f8bf547dd38",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist get gets multiple payloads 1`] = `
[
  {
    "nonce": 1,
    "schema": "network.xyo.debug",
  },
  {
    "nonce": 2,
    "schema": "network.xyo.test",
  },
  {
    "nonce": 3,
    "schema": "network.xyo.debug",
  },
]
`;

exports[`DeterministicArchivist get gets single boundwitness 1`] = `
[
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist get gets single payload 1`] = `
[
  {
    "nonce": 1,
    "schema": "network.xyo.debug",
  },
]
`;

exports[`DeterministicArchivist insert inserts multiple payloads 1`] = `
[
  {
    "addresses": [
      "3f1f91a4b4acb7d257c2be41b21cc9870509a77c",
      "ebeb7e68309874380df3e383084ebb1f1c60674e",
    ],
    "payload_hashes": [
      "ec0035a4a212de1f0028b5f1d4addff3fd7ba82b5b4de8ef5e7d096c0141a578",
    ],
    "payload_schemas": [
      "network.xyo.boundwitness",
    ],
    "previous_hashes": [
      "fa1e9d122274639bbe37342ef59a60f8de05fe3c37194bcfa8db7aaceb6d4f10",
      "fa1e9d122274639bbe37342ef59a60f8de05fe3c37194bcfa8db7aaceb6d4f10",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789002,
  },
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "1137543b5805a18e8f39cb7debe49ddf58a1de00c46ed7310baf7ebcfdf435a2",
      "aa36451ef0079af6cc2f86eaa05c7a93fb3c92098e18d0ce857e9e3072b32a85",
    ],
    "payload_schemas": [
      "network.xyo.debug",
      "network.xyo.test",
    ],
    "previous_hashes": [
      "d45ace1ac2f9e5d04dbf295850d47427ecd43f2e043c4a8326db2f8bf547dd38",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist insert inserts single payload 1`] = `
[
  {
    "addresses": [
      "3f1f91a4b4acb7d257c2be41b21cc9870509a77c",
      "ebeb7e68309874380df3e383084ebb1f1c60674e",
    ],
    "payload_hashes": [
      "d2fbc2df7cd956a2a9461792c9ffad4faf2ec290743832ff913582453e0dd320",
    ],
    "payload_schemas": [
      "network.xyo.boundwitness",
    ],
    "previous_hashes": [
      null,
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "addresses": [
      "db8af7b3084f5a03589af4915e37cbeb3f19afb0",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;
